### PR TITLE
Chain ID included in transaction ID's and predicate owners

### PR DESCRIPTION
**Pull request by Voxelot**
related to: https://github.com/FuelLabs/fuel-specs/pull/462, [#372](https://github.com/FuelLabs/fuel-vm/issues/372)

- hashes chain id into predicate owner
- hashes chain id into transaction id
- add GM op for fetching the chain id in the vm

### CODE REVIEW

File: fuel-tx/src/builder.rs [hunk 1]
```diff
@@ -276,25 +276,25 @@ impl<Tx: Buildable> TransactionBuilder<Tx> {
     }
 
     #[cfg(feature = "std")]
-    pub fn _finalize(&mut self) -> Tx {
+    pub fn _finalize(&mut self, parameters: &ConsensusParameters) -> Tx {
```
Comment (xgreenx) [inline fuel-tx/src/builder.rs:279]: It is better if the builder stores the `ConsensusParameters` inside. In this case, we don't need to propagate the parameter everywhere and break the API. 

We can use `Builder.set_consensus_parametrs` in tests where we want to use it. The same pattern that we use in the `fuel-core` where we have `Option<ConsensusParameters>` field=)

File: fuel-tx/src/transaction/types/input.rs [hunk 2]
```diff
@@ -566,23 +566,29 @@ impl Input {
         compute_message_id(sender, recipient, nonce, amount, data)
     }
 
-    pub fn predicate_owner<P>(predicate: P) -> Address
+    pub fn predicate_owner<P>(predicate: P, params: &ConsensusParameters) -> Address
```
Comment (xgreenx) [inline fuel-tx/src/transaction/types/input.rs:569]: Hmm, maybe we need to notify other teams about that explicitly

File: fuel-tx/src/transaction/types/script.rs [hunk 3]
```diff
@@ -102,7 +103,12 @@ impl crate::UniqueIdentifier for Script {
         clone.outputs_mut().iter_mut().for_each(Output::prepare_sign);
         clone.witnesses_mut().clear();
 
-        fuel_crypto::Hasher::hash(clone.to_bytes().as_slice())
+        let mut hasher = Hasher::default();
+        // chain ID
+        hasher.input(params.chain_id.to_be_bytes());
+        // transaction bytes
+        hasher.input(clone.to_bytes().as_slice());
+        hasher.finalize()
```
Comment (xgreenx) [inline fuel-tx/src/transaction/types/script.rs:111]: [nit] It Would be nice to move this code in the `compute_tx_id` function and reuse in 3 places=)

**Review by xgreenx — DISMISSED** [review]

File: fuel-tx/src/transaction/types/input.rs (hunk 2 above)
Comment (Voxelot) [reply fuel-tx/src/transaction/types/input.rs:569]: yep - it's a big change

**Review by xgreenx — APPROVED** [review]

**Pull request merged**